feature/wdelo-interactable

wdelo opened the pull request:
> :warning: NOTE: This is very barebones, will eventually need integration with the player pawn, inventory, and billboard
### Problem:

We do not have a base system in place for interactable objects in our game yet. This includes items, tools, and images, but will eventually include doors, evidence, lights, etc.

### Solution:

- Added the high level interface of `IInteractable`
- Added high level class of `Item`
- Added `Tool` and `Image` classes
- For each of above, I took the blueprints and added all of the functions and variables, filling in some functionality and adding comments where functionality will be needed.

### Additional Notes:

- If there are design concerns please let me know. I'm not entirely sure about the implementation design of what I've done so far, so it is always up to change. This is especially true since we haven't started to build on top of this, so now is a good time to change things with the base.
- I haven't worked with Unity's equivalent of Events or whatever Coroutines are so if we can do things better using those we probably should.

### Testing:

There is really nothing to test. It compiles and I haven't added any GameObjects into the world that use any of these scripts.

wdelo committed:
Added Interactable interface and Item, Tool, and Image classes.

Also added some skeleton functionality to each of the scripts I added.

wdelo commented:
Also why are there so many meta files ?

bossley9 commented on Assets/Scripts/Interactables/IInteractable.cs:
```diff
@@ -0,0 +1,15 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
```
This is a minor gripe but if we're not using namespaces we don't need to include them! It doesn't necessarily hurt the built game since it will all be compiled away (thanks to modern compilers) but including unused namespaces can potentially [slow down your editor (VS Code) and Unity compilation/lsp](https://stackoverflow.com/questions/1162766/how-is-performance-affected-by-an-unused-using-directive).

In VS Code if you have Mono/Unity lsp integration set up correctly, it should highlight which namespaces are being used and gray out which ones are not being used.
![VS Code graying out the using System namespaces](https://user-images.githubusercontent.com/35381535/112404823-78fb4800-8ce7-11eb-92d3-eb855d39f354.png)

bossley9 commented on Assets/Scripts/Interactables/IInteractable.cs:
```diff
@@ -0,0 +1,15 @@
+
+        public void OnStartInteract(Pawn interacter, Vector3 hitLocation, Vector3 hitNormal);
+
+        public void OnStopInteract();
```
I can't say too much since I'm not entirely familiar with the blueprints but you might want to look into events/delegates?
https://www.youtube.com/watch?v=VgwP1uXNnCw

bossley9 commented on Assets/Scripts/Interactables/Items/Image.cs:
```diff
@@ -0,0 +1,37 @@
+                // if YES item in hand is image, check if we are pinned 
+                //// if YES we are pinned, add the image in hand to billboard
+                //// if NO we are not pinned, check if we are not pinned permanently
+                ////// if YES, base.OnStartInteract(interacter, hitLocation, hitNormal);
```
`//////`

bossley9 commented on Assets/Scripts/Interactables/Items/Item.cs:
```diff
@@ -0,0 +1,39 @@
+
+namespace MMC
+{
+    public class Item : MonoBehaviour, IInteractable
```
Does this need to extend from `MonoBehavior`? Generally if a script needs lifecycle events (`Start` or `Update`), or if it will be attached as a script component to some game object, it should extend from `MonoBehavior`.

(I'm not saying you're right or wrong in extending - just mentioning for future reference)

bossley9 commented on Assets/Scripts/Interactables/Items/Item.cs:
```diff
@@ -0,0 +1,39 @@
+        public bool bIsInInventory;
+        public bool bIsInHand;
+        public Transform tInHandTransform;
+        public Transform tAimTransform;
```
Do all of these need to be public? The default access specifier is private. Remember, if something is public, it means that it can be accessed from the script component in the inspector.

![item public properties displaying graphically in the inspector window](https://user-images.githubusercontent.com/35381535/112405288-603f6200-8ce8-11eb-8558-6a13d3d8d7c8.png)

If you need special rules for accessing variables but you don't want them public in the inspector, you create getters and setters:
```cs
bool bIsInInventory; // don't need to specify "private" since this is default
...
public bool getIsInInventory()
{
    return bIsInInventory;
}
protected void SetIsInInventory(bool bNewVal)
{
    bIsInInventory = bNewVal;
}
```
or you can look into shorthand accessors:
```cs
bool bIsInInventory { get; protected set; }
```

wdelo replied:
So the shorthand accessors are actually properties, which are not serializable by the editor (they don't show up in the inspector). I just made variables that shouldn't be changed in the editor into public properties. Stuff like the aim position/rotation and in hand position/rotation should only be changed in the editor so I made it a private field with a [Serializable] tag and also a public property getter. 

For now I made all of the setter properties public since the inventory and billboard modify the variables that have setter properties.

bossley9 reviewed (approved):
Looks good

wdelo committed:
Accidently named class Picture instead of Image.

wdelo committed:
Removed using namespaces that we currently aren't using for now.

wdelo committed:
Removed 'On' prefixes for functions that had it. They aren't delegates so it's good to not confuse ourselves.

wdelo committed:
Added get/set properties for Item fields. Serialized the in hand and aim positions and rotations.

Instead of using Transform which needs a Game Object's transform, I changed it so the editor can enter a Vector3 and Quaternion for the position and rotation.

wdelo committed:
Made Tool functions virtusl so they can be overridden.

wdelo committed:
Made public properties since the inspector doesn't serialize properties.

Made private fields serializable but with public property getters, e.g. in hand transform should be tweaked in editor but not changed anywhere else.

bossley9 commented on Assets/Scripts/Interactables/Items/Image.cs:
```diff
@@ -0,0 +1,39 @@
+        //  Declare reference to current billboard here
+        //
+
+        public bool IsPinned { get; set; } // specifically "pinned" to a board
```
If the variable is public, this is just equivalent to
```cs
public bool IsPinned;
```
You probably mean to do something like
```cs
bool IsPinned { public get; protected set; }
```
or something. Leaving the `public` keyword nullifies the shorthand accessor properties.

The only time public shorthand accessors might be helpful is when you need to perform multiple actions in a setter:
```cs
public bool IsPinned { get; set {
    bIsSetting = true;
    vSomeVector = . . .
    IsPinned = value;
} }
```

[This overview](https://codeasy.net/lesson/properties) might be helpful. Ditto for the other accessors :)

bossley9 commented on Assets/Scripts/Interactables/Items/Item.cs:
```diff
@@ -0,0 +1,52 @@
+            }
+        }
+        public virtual void StopInteract() { }
+        public virtual void Select()
+        {
+            IsInHand = true;
+            gameObject.GetComponent<Renderer>().enabled = true;
+        }
+
+        public virtual void Deselect()
+        {
+            IsInHand = false;
+            gameObject.GetComponent<Renderer>().enabled = false;
+        }
```
I know we're just copying code so I don't expect any of this code to be optimized but I'll leave comments on Github for future reference since they'll be immortalized (oh no!).

Since `IsInHand` is always synced to the renderer component, we should make it reflect that way in the code:
```cs
void SetSelect(bool bIsSelected)
{
            IsInHand = bIsSelected;
            gameObject.GetComponent<Renderer>().enabled = IsInHand;
}
public virtual void Select()
{
            SetSelect(true);
}
public virtual void Deselect()
{
            SetSelect(false);
}
```
This is by no means a perfect solution but it creates a single source of truth. In this case, everything (and I mean _everything_) is synced uniformly with the variable `IsInHand`. Regardless of what it's set to, it's only set at one place. I also made `gameObject.GetComponent<Renderer>().enabled` set to the value because we can verify that they will always be the exact same value. This prevents typos or user error where someone might accidentally commit:
```cs
            IsInHand = true;
            gameObject.GetComponent<Renderer>().enabled = false;
```
We want to avoid user error if possible!

wdelo committed:
Made item selection functions more concise.

wdelo committed:
Made propertie setters to be protected. Also in the last commit I forgot to say that I made the Item class abstract.

wdelo committed:
Removed bIsInInventory field from Item

bossley9 commented on Assets/Scripts/Interactables/Items/Item.cs:
```diff
@@ -0,0 +1,54 @@
+                // if there is space then place item into inventory
+
+            }
+        }
```
You could do:
```cs
        public virtual void StartInteract(PlayerPawn player, Vector3 hitLocation, Vector3 hitNormal)
        {
                // check for space in the player inventory
                // if there is space then place item into inventory
        }
        public virtual void StartInteract(Pawn interacter, Vector3 hitLocation, Vector3 hitNormal)
        {
            PlayerPawn player = interacter as PlayerPawn;
            if (player != null)
                    StartInteract(player, hitLocation, hitNormal);
        }
```

wdelo committed:
Changed OnStartInteract parameter from Pawn to PlayerPawn
